File: lessons/tricks/reversible-compaction.ts
/**
 * Trick R: Reversible Compaction
 *
 * Preserves "reconstruction recipes" during context compaction.
 * Instead of discarding information, store retrieval keys so the model
 * can fetch details when needed.
 *
 * Problem: Traditional compaction loses important details forever.
 * When a file path or URL is summarized away, the model can't recover it.
 *
 * Solution: Extract and preserve references (URLs, file paths, function names)
 * as "reconstruction recipes" - minimal keys that enable re-retrieval.
 *
 * @example
 * ```typescript
 * import { createReversibleCompactor, extractReferences } from './reversible-compaction';
 *
 * const compactor = createReversibleCompactor({
 *   preserveTypes: ['file', 'url', 'function', 'error'],
 *   maxReferences: 50,
 * });
 *
 * // Compact with reference preservation
 * const result = compactor.compact(messages, {
 *   summarize: async (msgs) => summarizeLLM(msgs),
 * });
 *
 * console.log(result.summary);       // Condensed summary
 * console.log(result.references);    // Preserved references for retrieval
 * ```
 */

// =============================================================================
// TYPES
// =============================================================================

/**
 * Types of references that can be preserved.
 */
export type ReferenceType =
  | 'file'       // File paths
  | 'url'        // URLs (docs, issues, etc.)
  | 'function'   // Function/method names
  | 'class'      // Class/type names
  | 'error'      // Error messages/stack traces
  | 'command'    // Shell commands
  | 'snippet'    // Code snippets (abbreviated)
  | 'decision'   // Key decisions made
  | 'custom';    // Custom reference type

/**
 * A preserved reference that can be used for retrieval.
 */
export interface Reference {
  /** Unique ID for this reference */
  id: string;

  /** Type of reference */
  type: ReferenceType;

  /** The actual value (path, URL, name, etc.) */
  value: string;

  /** Brief context about this reference */
  context?: string;

  /** When this reference was encountered */
  timestamp: string;

  /** Source message index (before compaction) */
  sourceIndex?: number;

  /** Relevance score (0-1) for prioritization */
  relevance?: number;
}

/**
 * Configuration for reversible compaction.
 */
export interface ReversibleCompactionConfig {
  /** Types of references to preserve */
  preserveTypes: ReferenceType[];

  /** Maximum references to keep */
  maxReferences?: number;

  /** Whether to deduplicate references */
  deduplicate?: boolean;

  /** Custom extractors for specific reference types */
  customExtractors?: Map<ReferenceType, ReferenceExtractor>;

  /** Minimum relevance score to preserve (0-1) */
  minRelevance?: number;
}

/**
 * Result of compaction.
 */
export interface CompactionResult {
  /** The summarized content */
  summary: string;

  /** Preserved references for retrieval */
  references: Reference[];

  /** Statistics about the compaction */
  stats: CompactionStats;
}

/**
 * Statistics about compaction.
 */
export interface CompactionStats {
  /** Original message count */
  originalMessages: number;

  /** Original estimated tokens */
  originalTokens: number;

  /** Compacted token count */
  compactedTokens: number;

  /** References extracted */
  referencesExtracted: number;

  /** References preserved (after dedup/limits) */
  referencesPreserved: number;

  /** Compression ratio */
  compressionRatio: number;
}

/**
 * Message for compaction.
 */
export interface CompactionMessage {
  role: 'system' | 'user' | 'assistant' | 'tool';
  content: string;
  name?: string;
}

/**
 * Custom reference extractor function.
 */
export type ReferenceExtractor = (content: string, index: number) => Reference[];

/**
 * Options for compact operation.
 */
export interface CompactOptions {
  /** Function to summarize messages */
  summarize: (messages: CompactionMessage[]) => Promise<string>;

  /** Additional context to preserve */
  additionalContext?: string;
}

/**
 * Events emitted during compaction.
 */
export type CompactionEvent =
  | { type: 'compaction.started'; messageCount: number }
  | { type: 'reference.extracted'; reference: Reference }
  | { type: 'reference.deduplicated'; kept: number; removed: number }
  | { type: 'compaction.completed'; stats: CompactionStats };

export type CompactionEventListener = (event: CompactionEvent) => void;

// =============================================================================
// REFERENCE EXTRACTION
// =============================================================================

/**
 * Extract file path references from content.
 */
export function extractFileReferences(content: string, sourceIndex?: number): Reference[] {
  const references: Reference[] = [];
  const timestamp = new Date().toISOString();

  // Unix-style paths
  const unixPaths = content.match(/(?:\/[\w.-]+)+(?:\/[\w.-]*)?/g) || [];

  // Windows-style paths
  const windowsPaths = content.match(/[A-Za-z]:\\(?:[\w.-]+\\)*[\w.-]*/g) || [];

  // Relative paths with extensions
  const relativePaths = content.match(/(?:\.\.?\/)?[\w-]+(?:\/[\w.-]+)*\.\w+/g) || [];

  const allPaths = [...new Set([...unixPaths, ...windowsPaths, ...relativePaths])];

  for (const path of allPaths) {
    // Filter out obvious non-paths
    if (path.length < 3 || path === '/' || path.match(/^\/\d+$/)) continue;

    references.push({
      id: `file-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
      type: 'file',
      value: path,
      timestamp,
      sourceIndex,
    });
  }

  return references;
}

/**
 * Extract URL references from content.
 */
export function extractUrlReferences(content: string, sourceIndex?: number): Reference[] {
  const references: Reference[] = [];
  const timestamp = new Date().toISOString();

  // Full URLs
  const urlPattern = /https?:\/\/[^\s<>"{}|\\^`\[\]]+/g;
  const urls = content.match(urlPattern) || [];

  for (const url of urls) {
    // Clean trailing punctuation
    const cleanUrl = url.replace(/[.,;:!?)]+$/, '');

    // Extract context from URL
    let context: string | undefined;
    if (cleanUrl.includes('github.com')) {
      context = 'GitHub';
      if (cleanUrl.includes('/issues/')) context += ' Issue';
      if (cleanUrl.includes('/pull/')) context += ' PR';
    } else if (cleanUrl.includes('stackoverflow.com')) {
      context = 'Stack Overflow';
    } else if (cleanUrl.includes('docs.')) {
      context = 'Documentation';
    }

    references.push({
      id: `url-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
      type: 'url',
      value: cleanUrl,
      context,
      timestamp,
      sourceIndex,
    });
  }

  return references;
}

/**
 * Extract function/method references from content.
 */
export function extractFunctionReferences(content: string, sourceIndex?: number): Reference[] {
  const references: Reference[] = [];
  const timestamp = new Date().toISOString();

  // Function definitions
  const funcDefs = content.match(/(?:function|async function|const|let|var)\s+(\w+)\s*(?:=\s*(?:async\s*)?\([^)]*\)\s*=>|\([^)]*\))/g) || [];

  // Method calls with meaningful names
  const methodCalls = content.match(/\b([a-z][a-zA-Z0-9]*(?:[A-Z][a-zA-Z0-9]*)+)\s*\(/g) || [];

  const seen = new Set<string>();

  for (const match of funcDefs) {
    const name = match.match(/(?:function|const|let|var)\s+(\w+)/)?.[1];
    if (name && !seen.has(name) && name.length > 2) {
      seen.add(name);
      references.push({
        id: `func-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
        type: 'function',
        value: name,
        context: 'definition',
        timestamp,
        sourceIndex,
      });
    }
  }

  for (const match of methodCalls) {
    const name = match.replace(/\s*\($/, '');
    if (!seen.has(name) && name.length > 2) {
      seen.add(name);
      references.push({
        id: `func-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
        type: 'function',
        value: name,
        context: 'call',
        timestamp,
        sourceIndex,
      });
    }
  }

  return references;
}

/**
 * Extract error references from content.
 */
export function extractErrorReferences(content: string, sourceIndex?: number): Reference[] {
  const references: Reference[] = [];
  const timestamp = new Date().toISOString();

  // Error class names
  const errorClasses = content.match(/\b(\w+Error|\w+Exception)\b/g) || [];

  // Error messages (e.g., "Error: something went wrong")
  const errorMessages = content.match(/(?:Error|Exception|Failed|Failure):\s*[^\n]+/g) || [];

  const seen = new Set<string>();

  for (const errorClass of errorClasses) {
    if (!seen.has(errorClass)) {
      seen.add(errorClass);
      references.push({
        id: `error-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
        type: 'error',
        value: errorClass,
        context: 'error type',
        timestamp,
        sourceIndex,
      });
    }
  }

  for (const errorMsg of errorMessages.slice(0, 3)) { // Limit error messages
    const truncated = errorMsg.slice(0, 100);
    if (!seen.has(truncated)) {
      seen.add(truncated);
      references.push({
        id: `error-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
        type: 'error',
        value: truncated,
        context: 'error message',
        timestamp,
        sourceIndex,
      });
    }
  }

  return references;
}

/**
 * Extract command references from content.
 */
export function extractCommandReferences(content: string, sourceIndex?: number): Reference[] {
  const references: Reference[] = [];
  const timestamp = new Date().toISOString();

  // Shell commands (common patterns)
  const shellPatterns = [
    /\$\s*([^\n]+)/g,           // $ command
    /```(?:bash|sh|shell)\n([^`]+)```/g,  // Code blocks
    /(?:npm|yarn|pnpm|npx|bun)\s+[^\n]+/g,  // Package managers
    /(?:git)\s+[^\n]+/g,        // Git commands
    /(?:docker|kubectl)\s+[^\n]+/g,  // Container commands
  ];

  const seen = new Set<string>();

  for (const pattern of shellPatterns) {
    const matches = content.match(pattern) || [];
    for (const match of matches) {
      const cleaned = match.replace(/^\$\s*/, '').replace(/```(?:bash|sh|shell)?\n?/g, '').trim();
      if (cleaned.length > 3 && !seen.has(cleaned)) {
        seen.add(cleaned);
        references.push({
          id: `cmd-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
          type: 'command',
          value: cleaned.slice(0, 200), // Truncate long commands
          timestamp,
          sourceIndex,
        });
      }
    }
  }

  return references;
}

/**
 * Extract all references from content.
 */
export function extractReferences(
  content: string,
  types: ReferenceType[],
  sourceIndex?: number,
  customExtractors?: Map<ReferenceType, ReferenceExtractor>
): Reference[] {
  const references: Reference[] = [];

  for (const type of types) {
    // Check for custom extractor first
    if (customExtractors?.has(type)) {
      const extractor = customExtractors.get(type)!;
      references.push(...extractor(content, sourceIndex ?? 0));
      continue;
    }

    // Use built-in extractors
    switch (type) {
      case 'file':
        references.push(...extractFileReferences(content, sourceIndex));
        break;
      case 'url':
        references.push(...extractUrlReferences(content, sourceIndex));
        break;
      case 'function':
        references.push(...extractFunctionReferences(content, sourceIndex));
        break;
      case 'error':
        references.push(...extractErrorReferences(content, sourceIndex));
        break;
      case 'command':
        references.push(...extractCommandReferences(content, sourceIndex));
        break;
      // class, snippet, decision, custom require custom extractors
    }
  }

  return references;
}

// =============================================================================
// REVERSIBLE COMPACTOR
// =============================================================================

/**
 * Manages reversible context compaction.
 */
export class ReversibleCompactor {
  private config: Required<ReversibleCompactionConfig>;
  private preservedReferences: Reference[] = [];
  private listeners: CompactionEventListener[] = [];

  constructor(config: ReversibleCompactionConfig) {
    this.config = {
      preserveTypes: config.preserveTypes,
      maxReferences: config.maxReferences ?? 100,
      deduplicate: config.deduplicate ?? true,
      customExtractors: config.customExtractors ?? new Map(),
      minRelevance: config.minRelevance ?? 0,
    };
  }

  /**
   * Compact messages while preserving references.
   */
  async compact(
    messages: CompactionMessage[],
    options: CompactOptions
  ): Promise<CompactionResult> {
    this.emit({ type: 'compaction.started', messageCount: messages.length });

    // Calculate original token estimate
    const originalContent = messages.map(m => m.content).join('\n');
    const originalTokens = Math.ceil(originalContent.length / 4);

    // Extract references from all messages
    const allReferences: Reference[] = [];
    for (let i = 0; i < messages.length; i++) {
      const refs = extractReferences(
        messages[i].content,
        this.config.preserveTypes,
        i,
        this.config.customExtractors
      );

      for (const ref of refs) {
        this.emit({ type: 'reference.extracted', reference: ref });
        allReferences.push(ref);
      }
    }

    // Deduplicate if enabled
    let processedReferences = allReferences;
    if (this.config.deduplicate) {
      processedReferences = this.deduplicateReferences(allReferences);
    }

    // Filter by relevance
    if (this.config.minRelevance > 0) {
      processedReferences = processedReferences.filter(
        r => (r.relevance ?? 1) >= this.config.minRelevance
      );
    }

    // Limit references
    if (processedReferences.length > this.config.maxReferences) {
      // Sort by relevance, keep most relevant
      processedReferences.sort((a, b) => (b.relevance ?? 0.5) - (a.relevance ?? 0.5));
      processedReferences = processedReferences.slice(0, this.config.maxReferences);
    }

    // Generate summary
    const summary = await options.summarize(messages);

    // Calculate compacted token estimate
    const referenceBlock = this.formatReferencesBlock(processedReferences);
    const compactedContent = summary + '\n\n' + referenceBlock;
    const compactedTokens = Math.ceil(compactedContent.length / 4);

    // Store preserved references
    this.preservedReferences = processedReferences;

    const stats: CompactionStats = {
      originalMessages: messages.length,
      originalTokens,
      compactedTokens,
      referencesExtracted: allReferences.length,
      referencesPreserved: processedReferences.length,
      compressionRatio: compactedTokens / originalTokens,
    };

    this.emit({ type: 'compaction.completed', stats });

    return {
      summary,
      references: processedReferences,
      stats,
    };
  }

  /**
   * Format references as a block for inclusion in context.
   */
  formatReferencesBlock(references: Reference[]): string {
    if (references.length === 0) return '';

    const grouped = new Map<ReferenceType, Reference[]>();
    for (const ref of references) {
      const group = grouped.get(ref.type) || [];
      group.push(ref);
      grouped.set(ref.type, group);
    }

    const parts: string[] = ['[Preserved References]'];

    for (const [type, refs] of grouped) {
      parts.push(`\n${type.toUpperCase()}S:`);
      for (const ref of refs) {
        const contextPart = ref.context ? ` (${ref.context})` : '';
        parts.push(`  - ${ref.value}${contextPart}`);
      }
    }

    return parts.join('\n');
  }

  /**
   * Get a reference by ID.
   */
  getReference(id: string): Reference | undefined {
    return this.preservedReferences.find(r => r.id === id);
  }

  /**
   * Get references by type.
   */
  getReferencesByType(type: ReferenceType): Reference[] {
    return this.preservedReferences.filter(r => r.type === type);
  }

  /**
   * Search references by value.
   */
  searchReferences(query: string): Reference[] {
    const lowerQuery = query.toLowerCase();
    return this.preservedReferences.filter(
      r => r.value.toLowerCase().includes(lowerQuery)
    );
  }

  /**
   * Get all preserved references.
   */
  getPreservedReferences(): Reference[] {
    return [...this.preservedReferences];
  }

  /**
   * Clear preserved references.
   */
  clear(): void {
    this.preservedReferences = [];
  }

  /**
   * Subscribe to events.
   */
  on(listener: CompactionEventListener): () => void {
    this.listeners.push(listener);
    return () => {
      const idx = this.listeners.indexOf(listener);
      if (idx >= 0) this.listeners.splice(idx, 1);
    };
  }

  // Internal methods

  private deduplicateReferences(references: Reference[]): Reference[] {
    const seen = new Map<string, Reference>();
    let removed = 0;

    for (const ref of references) {
      const key = `${ref.type}:${ref.value}`;
      if (!seen.has(key)) {
        seen.set(key, ref);
      } else {
        removed++;
      }
    }

    this.emit({
      type: 'reference.deduplicated',
      kept: seen.size,
      removed,
    });

    return Array.from(seen.values());
  }

  private emit(event: CompactionEvent): void {
    for (const listener of this.listeners) {
      try {
        listener(event);
      } catch {
        // Ignore listener errors
      }
    }
  }
}

// =============================================================================
// FACTORY FUNCTIONS
// =============================================================================

/**
 * Create a reversible compactor.
 *
 * @example
 * ```typescript
 * const compactor = createReversibleCompactor({
 *   preserveTypes: ['file', 'url', 'error'],
 *   maxReferences: 50,
 *   deduplicate: true,
 * });
 *
 * const result = await compactor.compact(messages, {
 *   summarize: async (msgs) => {
 *     // Use LLM to summarize
 *     return await llm.summarize(msgs);
 *   },
 * });
 *
 * // Result includes both summary and preserved references
 * console.log(result.summary);
 * console.log(result.references);
 *
 * // Later, search for specific references
 * const fileRefs = compactor.getReferencesByType('file');
 * const errorRefs = compactor.searchReferences('TypeError');
 * ```
 */
export function createReversibleCompactor(
  config: ReversibleCompactionConfig
): ReversibleCompactor {
  return new ReversibleCompactor(config);
}

// =============================================================================
// UTILITIES
// =============================================================================

/**
 * Quick helper to extract references without full compaction.
 */
export function quickExtract(
  content: string,
  types: ReferenceType[] = ['file', 'url', 'function', 'error']
): Reference[] {
  return extractReferences(content, types);
}

/**
 * Create a reconstruction prompt for re-fetching details.
 */
export function createReconstructionPrompt(references: Reference[]): string {
  if (references.length === 0) {
    return '';
  }

  const lines = [
    'The following references were preserved from earlier context.',
    'Use these to retrieve details if needed:',
    '',
  ];

  const byType = new Map<ReferenceType, Reference[]>();
  for (const ref of references) {
    const group = byType.get(ref.type) || [];
    group.push(ref);
    byType.set(ref.type, group);
  }

  if (byType.has('file')) {
    lines.push('**Files** (can be read with read_file tool):');
    for (const ref of byType.get('file')!) {
      lines.push(`  - ${ref.value}`);
    }
    lines.push('');
  }

  if (byType.has('url')) {
    lines.push('**URLs** (can be fetched for details):');
    for (const ref of byType.get('url')!) {
      const ctx = ref.context ? ` [${ref.context}]` : '';
      lines.push(`  - ${ref.value}${ctx}`);
    }
    lines.push('');
  }

  if (byType.has('function')) {
    lines.push('**Functions** (search codebase if details needed):');
    for (const ref of byType.get('function')!) {
      lines.push(`  - ${ref.value}`);
    }
    lines.push('');
  }

  if (byType.has('error')) {
    lines.push('**Errors encountered**:');
    for (const ref of byType.get('error')!) {
      lines.push(`  - ${ref.value}`);
    }
    lines.push('');
  }

  if (byType.has('command')) {
    lines.push('**Commands used**:');
    for (const ref of byType.get('command')!) {
      lines.push(`  - ${ref.value}`);
    }
  }

  return lines.join('\n');
}

/**
 * Calculate relevance score for a reference based on context.
 */
export function calculateRelevance(
  reference: Reference,
  context: { goal?: string; recentTopics?: string[] }
): number {
  let score = 0.5; // Base score

  const value = reference.value.toLowerCase();

  // Boost if matches goal
  if (context.goal) {
    const goalWords = context.goal.toLowerCase().split(/\s+/);
    for (const word of goalWords) {
      if (word.length > 3 && value.includes(word)) {
        score += 0.1;
      }
    }
  }

  // Boost if matches recent topics
  if (context.recentTopics) {
    for (const topic of context.recentTopics) {
      if (value.includes(topic.toLowerCase())) {
        score += 0.15;
      }
    }
  }

  // Type-based adjustments
  if (reference.type === 'error') score += 0.1; // Errors are usually important
  if (reference.type === 'file') score += 0.05; // Files are often needed

  return Math.min(1, score);
}

/**
 * Format compaction stats for display.
 */
export function formatCompactionStats(stats: CompactionStats): string {
  const compressionPercent = Math.round((1 - stats.compressionRatio) * 100);

  return `Compaction Statistics:
  Original:    ${stats.originalMessages} messages, ~${stats.originalTokens.toLocaleString()} tokens
  Compacted:   ~${stats.compactedTokens.toLocaleString()} tokens
  Compression: ${compressionPercent}% reduction
  References:  ${stats.referencesExtracted} extracted, ${stats.referencesPreserved} preserved`;
}

File: attocode/attocode_py/src/attocode/tricks/reversible_compaction.py
"""Reversible compaction with reference preservation (Trick R).

Compresses conversation history while preserving references to
files, URLs, functions, errors, and commands for later retrieval.
"""

from __future__ import annotations

import re
import time
import uuid
from dataclasses import dataclass, field
from typing import Any, Callable

from attocode.integrations.utilities.token_estimate import estimate_tokens


class ReferenceType:
    """Types of extractable references."""

    FILE = "file"
    URL = "url"
    FUNCTION = "function"
    CLASS = "class"
    ERROR = "error"
    COMMAND = "command"
    SNIPPET = "snippet"
    DECISION = "decision"
    CUSTOM = "custom"


@dataclass
class Reference:
    """A preserved reference from compacted content."""

    id: str
    type: str
    value: str
    context: str | None = None
    timestamp: float = field(default_factory=time.time)
    source_index: int | None = None
    relevance: float = 0.5


@dataclass
class CompactionStats:
    """Statistics about a compaction operation."""

    original_messages: int = 0
    original_tokens: int = 0
    compacted_tokens: int = 0
    references_extracted: int = 0
    references_preserved: int = 0
    compression_ratio: float = 0.0


@dataclass
class CompactionResult:
    """Result of a compaction operation."""

    summary: str
    references: list[Reference]
    stats: CompactionStats


@dataclass
class ReversibleCompactionConfig:
    """Configuration for reversible compaction."""

    preserve_types: list[str] = field(
        default_factory=lambda: ["file", "url", "function", "error"]
    )
    max_references: int = 100
    deduplicate: bool = True
    min_relevance: float = 0.0


ReferenceExtractor = Callable[[str, int], list[Reference]]
CompactionEventListener = Callable[[str, dict[str, Any]], None]


class ReversibleCompactor:
    """Compacts conversation while preserving key references.

    Extracts file paths, URLs, function names, errors, and commands
    from messages before summarization. References can be retrieved
    later for context reconstruction.
    """

    def __init__(self, config: ReversibleCompactionConfig | None = None) -> None:
        self._config = config or ReversibleCompactionConfig()
        self._references: list[Reference] = []
        self._listeners: list[CompactionEventListener] = []
        self._custom_extractors: dict[str, ReferenceExtractor] = {}

    async def compact(
        self,
        messages: list[dict[str, Any]],
        summarize: Callable[[list[dict[str, Any]]], Any],
        additional_context: str | None = None,
    ) -> CompactionResult:
        """Compact messages with reference preservation.

        Args:
            messages: Messages to compact.
            summarize: Async function that summarizes messages into a string.
            additional_context: Extra context for the summary.
        """
        # Extract references from all messages
        all_refs: list[Reference] = []
        original_tokens = 0

        for i, msg in enumerate(messages):
            content = msg.get("content", "")
            if isinstance(content, str):
                original_tokens += estimate_tokens(content)
                refs = extract_references(
                    content, self._config.preserve_types, i, self._custom_extractors
                )
                all_refs.extend(refs)

        # Deduplicate
        if self._config.deduplicate:
            seen: set[str] = set()
            deduped: list[Reference] = []
            for ref in all_refs:
                key = f"{ref.type}:{ref.value}"
                if key not in seen:
                    seen.add(key)
                    deduped.append(ref)
            all_refs = deduped

        # Filter by relevance
        if self._config.min_relevance > 0:
            all_refs = [r for r in all_refs if r.relevance >= self._config.min_relevance]

        # Enforce max references (keep highest relevance)
        if len(all_refs) > self._config.max_references:
            all_refs.sort(key=lambda r: r.relevance, reverse=True)
            all_refs = all_refs[: self._config.max_references]

        # Store references
        self._references.extend(all_refs)

        # Get summary
        result = summarize(messages)
        if hasattr(result, "__await__"):
            summary = await result
        else:
            summary = result

        compacted_tokens = estimate_tokens(str(summary))

        stats = CompactionStats(
            original_messages=len(messages),
            original_tokens=original_tokens,
            compacted_tokens=compacted_tokens,
            references_extracted=len(all_refs),
            references_preserved=len(all_refs),
            compression_ratio=compacted_tokens / max(1, original_tokens),
        )

        self._emit("compaction.completed", {"stats": stats})

        return CompactionResult(
            summary=str(summary),
            references=all_refs,
            stats=stats,
        )

    def format_references_block(self, references: list[Reference] | None = None) -> str:
        """Format references as a text block for context injection."""
        refs = references or self._references
        if not refs:
            return ""

        grouped: dict[str, list[str]] = {}
        for ref in refs:
            label = ref.type.upper() + "S"
            if label not in grouped:
                grouped[label] = []
            grouped[label].append(f"  - {ref.value}")

        lines = ["[Preserved References]"]
        for label, items in grouped.items():
            lines.append(f"{label}:")
            lines.extend(items)
        return "\n".join(lines)

    def get_reference(self, ref_id: str) -> Reference | None:
        """Get a reference by ID."""
        for ref in self._references:
            if ref.id == ref_id:
                return ref
        return None

    def get_references_by_type(self, ref_type: str) -> list[Reference]:
        """Get all references of a given type."""
        return [r for r in self._references if r.type == ref_type]

    def search_references(self, query: str) -> list[Reference]:
        """Search references by value."""
        query_lower = query.lower()
        return [r for r in self._references if query_lower in r.value.lower()]

    def get_preserved_references(self) -> list[Reference]:
        """Get all preserved references."""
        return list(self._references)

    def clear(self) -> None:
        """Clear all stored references."""
        self._references.clear()

    def on(self, listener: CompactionEventListener) -> Callable[[], None]:
        """Subscribe to compaction events."""
        self._listeners.append(listener)

        def unsubscribe() -> None:
            try:
                self._listeners.remove(listener)
            except ValueError:
                pass

        return unsubscribe

    def _emit(self, event: str, data: dict[str, Any]) -> None:
        for listener in self._listeners:
            try:
                listener(event, data)
            except Exception:
                pass


# --- Reference extractors ---

def extract_file_references(content: str, source_index: int = 0) -> list[Reference]:
    """Extract file path references."""
    refs: list[Reference] = []
    # Unix/Mac paths with extensions
    for m in re.finditer(r"(?:^|\s|['\"`(])(/[\w./-]+\.\w{1,6})\b", content):
        refs.append(_make_ref(ReferenceType.FILE, m.group(1).strip(), source_index))
    # Relative paths with extensions
    for m in re.finditer(r"(?:^|\s|['\"`(])(\.{0,2}/[\w./-]+\.\w{1,6})\b", content):
        val = m.group(1).strip()
        if not any(r.value == val for r in refs):
            refs.append(_make_ref(ReferenceType.FILE, val, source_index))
    return refs


def extract_url_references(content: str, source_index: int = 0) -> list[Reference]:
    """Extract URL references."""
    refs: list[Reference] = []
    for m in re.finditer(r"https?://[^\s<>\")]+", content):
        url = m.group(0).rstrip(".,;:")
        refs.append(_make_ref(ReferenceType.URL, url, source_index))
    return refs


def extract_function_references(content: str, source_index: int = 0) -> list[Reference]:
    """Extract function/method references."""
    refs: list[Reference] = []
    # Function definitions
    for m in re.finditer(r"(?:def|function|async\s+def)\s+(\w+)", content):
        refs.append(_make_ref(ReferenceType.FUNCTION, m.group(1), source_index))
    # camelCase method calls
    for m in re.finditer(r"\b([a-z][a-zA-Z]{2,})\s*\(", content):
        name = m.group(1)
        if len(name) > 2 and not any(r.value == name for r in refs):
            refs.append(_make_ref(ReferenceType.FUNCTION, name, source_index))
    return refs


def extract_error_references(content: str, source_index: int = 0) -> list[Reference]:
    """Extract error references."""
    refs: list[Reference] = []
    # Error class names
    for m in re.finditer(r"\b(\w+Error|\w+Exception)\b", content):
        if len(refs) < 3:
            refs.append(_make_ref(ReferenceType.ERROR, m.group(1), source_index))
    # Error messages
    for m in re.finditer(r"Error:\s*(.{10,80})", content):
        if len(refs) < 3:
            refs.append(_make_ref(ReferenceType.ERROR, m.group(0).strip(), source_index))
    return refs


def extract_command_references(content: str, source_index: int = 0) -> list[Reference]:
    """Extract command references."""
    refs: list[Reference] = []
    # $ prefix commands
    for m in re.finditer(r"^\$\s+(.+)$", content, re.MULTILINE):
        refs.append(_make_ref(ReferenceType.COMMAND, m.group(1).strip(), source_index))
    # Code block commands
    for m in re.finditer(r"```(?:bash|sh|shell)?\n(.+?)```", content, re.DOTALL):
        lines = m.group(1).strip().split("\n")
        for line in lines[:3]:
            line = line.strip()
            if line and not line.startswith("#"):
                refs.append(_make_ref(ReferenceType.COMMAND, line, source_index))
    # Common CLI patterns
    for m in re.finditer(r"\b((?:npm|git|docker|pip|python|node)\s+\S+(?:\s+\S+){0,3})", content):
        cmd = m.group(1).strip()
        if not any(r.value == cmd for r in refs):
            refs.append(_make_ref(ReferenceType.COMMAND, cmd, source_index))
    return refs


def extract_references(
    content: str,
    types: list[str],
    source_index: int = 0,
    custom_extractors: dict[str, ReferenceExtractor] | None = None,
) -> list[Reference]:
    """Extract all references of specified types from content."""
    _extractors: dict[str, Callable[[str, int], list[Reference]]] = {
        ReferenceType.FILE: extract_file_references,
        ReferenceType.URL: extract_url_references,
        ReferenceType.FUNCTION: extract_function_references,
        ReferenceType.ERROR: extract_error_references,
        ReferenceType.COMMAND: extract_command_references,
    }

    all_refs: list[Reference] = []
    for ref_type in types:
        extractor = (custom_extractors or {}).get(ref_type) or _extractors.get(ref_type)
        if extractor:
            all_refs.extend(extractor(content, source_index))
    return all_refs


def quick_extract(content: str, types: list[str] | None = None) -> list[Reference]:
    """Quick extraction with default types."""
    types = types or ["file", "url", "function", "error"]
    return extract_references(content, types)


def calculate_relevance(
    reference: Reference,
    goal: str | None = None,
    recent_topics: list[str] | None = None,
) -> float:
    """Calculate relevance score for a reference."""
    score = 0.5

    if goal:
        goal_words = {w.lower() for w in goal.split() if len(w) > 3}
        ref_words = {w.lower() for w in reference.value.split("/") + reference.value.split(".")}
        overlap = goal_words & ref_words
        score += len(overlap) * 0.1

    if recent_topics:
        for topic in recent_topics:
            if reference.value.lower() in topic.lower() or topic.lower() in reference.value.lower():
                score += 0.15

    if reference.type == ReferenceType.ERROR:
        score += 0.1
    elif reference.type == ReferenceType.FILE:
        score += 0.05

    return min(1.0, score)


def _make_ref(ref_type: str, value: str, source_index: int) -> Reference:
    """Create a reference with a unique ID."""
    return Reference(
        id=f"ref-{uuid.uuid4().hex[:8]}",
        type=ref_type,
        value=value,
        source_index=source_index,
    )
